### ISSUE TITLE
Does mgo do smart projections based on type of result?

### ISSUE THREAD

**Issue by JustinDrake** [go-mgo/mgo#483]
When a query is made without using [`Select`](https://godoc.org/gopkg.in/mgo.v2#Query.Select) is an implicit projection still made based on the type of the result object? (I'm filing this issue because the documentation is not super clear on this, and it's quite an important point.)

Consider this query:

```
type Test struct {
    Timestamp string `bson:"timestamp"`
}

var test Test
myCollection.Find(bson.M{
    "test": true,
}).One(&test)
```

The result `test` of type `Test` only has a single field `Timestamp`. I'm hoping the above is equivalent (in performance) to:

```
type Test struct {
    Timestamp string `bson:"timestamp"`
}

var test Test
myCollection.Find(bson.M{
    "test": true,
}).Select(bson.M{
    "timestamp": true,
}).One(&test)
```

In the above query I have an explicit (and redundant!) projection using `Select`. I am hoping that mgo does implicit projections, otherwise I will have to explicitly add `Select`s to avoid unnecessarily fetching data from the database.

**Comment by vivace-io** [comment]
I know the question is old now, but MongoDB would be the bottleneck in the code above. The `mgo` package primarily serves as a communication layer with a Mongo server. A second Go function call will (obviously) add overhead here, but the interpretation of input is still largely left to MongoDB. Think of `mgo` as an interpreter in this instance.

You *will* get worse performance with additional `mgo` calls, albeit minimal. But MongoDB itself is the deciding factor in performance between these two calls. 

In other words: changing between the two in your Go code will not change a noticeable amount to your end user, at scale or testing.
